### DISCUSSION TITLE
How to interpret sleep data?

### DISCUSSION THREAD

**Discussion by GrumpyMeow** [YorkshireIoT/ha-google-fit#201] [Q&A]
A while ago i installed this HACS integration. I have a Google Home Hub next to my bed which tracks me sleeping. I do see the data being registered by the integration. But for instance the value of "ligh-sleep" in the chart increases, but also decreases. It's like i'm undoing my light-sleeping :-S. I think it has got something to do with the data counting the is calculated over the last 24 hours. So probably the night before i woke-up, or this night i woke-up... It's hard to interpret the presented data.
I think i would like the data to be presented in the number of seconds i was in light-sleep, the number of seconds i was in deep-sleep, etc. 
How do you guys interpret the given data? Do you make use of some calculated-sensors?

![image](https://github.com/YorkshireIoT/ha-google-fit/assets/12073499/9d3633e1-7465-4a3d-9c3e-db499b1d30b8)

**Comment by YorkshireIoT** [comment]
Hi @GrumpyMeow,

The data from the integration is number of seconds in the last 24 hour period.

I think by your suggestion, you mean you'd like to see the number of seconds asleep in the last single sleep session. This is tricky, as depending on the data provider to Google Fit the given sleep session could actually be broken down into multiple sessions which would then lead to more confusing data reporting.

Please correct me if I've misinterpreted you

**Reply by MalteMatthey** [reply]
Hi @YorkshireIoT ,
until now I didn't dive deep into the code and tried to understand how google tracks the sleep data. Can you explain, what exactly the sleep in the last 24h period is measuring?

The accurate seconds in the last 24h would allow for our own calculations for the last sleep session.
You could do something like this, to calculate the sleep time from 8pm on:
`TimeSleptFrom8pmOn = (currentSleepValue - SleepValueAt8pm) - (SleepValue24hAgo - SleepValueAt8pm24hAgo)`
This formula would compensate that the sleep from the night before disappears after 24h.

However, the sleep seems to be tracked differently and I don't really understand how. Somehow the sleep from the whole session the night before dissapears in one big chunk. This makes it very hard to calculate anything. (Even when only calculation the positive difference and ignoring the dissapearing of the night before, I don't get an accurate value.) Can you explain, what the 24h value is really measuring? 

If my sleep is exactly the same every night the sensor value should be constant from my understanding, but this isn't the case. (When sleeping at the same time as the last night, the current sleep should cancel out the sleep from the last night that is lost because it dosn't belong to the last 24h period anymore)

Thanks for the great integration though!

**Reply by YorkshireIoT** [reply]
The line of code specifying the last 24 hours is [here](https://github.com/YorkshireIoT/ha-google-fit/blob/main/custom_components/google_fit/const.py#L189). This in then passed to `_get_interval` which uses it [here](https://github.com/YorkshireIoT/ha-google-fit/blob/main/custom_components/google_fit/coordinator.py#L111) to find the start time.

I'm not quite sure why the sleep time looks weird for some people. It's on my list of things to have a look at next though 👍

**Reply by MalteMatthey** [reply]
Thanks for the answer! I couldn't find the cause of the problem until now. If I have some spare time weeks, I can analyse it a bit further.

If that helps - for me only the sensor.sleep does not seem to work, deep sleep, light sleep, rem sleep and the awake time seem to be correct. I'll check if adding those 3 sensor values will give me an accurate sleep sensor and will report back in a few days!

![image](https://github.com/YorkshireIoT/ha-google-fit/assets/116491333/012b3773-aef9-4bf8-98c9-41fbe199c6b8)

The second graph are the deep, light and rem sleep values aggregated.

**Comment by GrumpyMeow** [comment]
Hi Yorkshire,
I indeed would like to get data for a single sleep cycle, which consists out of some light+deep+rem parts. I do understand the complexity it presents.
